student/src/main/java/tds/student/web/handlers/TestShellHandler.java

```diff
@@ -226,54 +231,24 @@ private String parseOutLatencies (Map<String, String> formParams) {
 
   @RequestMapping (value = "TestShell.axd/complete")
   @ResponseBody
-  public ResponseData<String> reviewTest (@RequestParam Map<String, String> formParams, HttpServletRequest request) throws IOException, TDSSecurityException, ReturnStatusException {
+  public ResponseData<String> reviewTest (@RequestParam Map<String, String> formParams, HttpServletRequest request)
+    throws IOException, TDSSecurityException, ReturnStatusException {
     checkAuthenticated ();
 
     TestOpportunity testOpp = StudentContext.getTestOpportunity ();
 
-    // get responses
-    TestManager testManager = new TestManager (testOpp);
-    testManager.LoadResponses (true);
+    ResponseData<String> responseData = reviewTestService.reviewTest(testOpp, new TestManager(testOpp));
 
-    // check if test length is met
-    testManager.CheckIfTestComplete ();
+    if (responseData.getReplyCode() == TDSReplyCode.Error.getCode()) {
```

Comment:
Should this be `if(responsedata.getReplyCode() != TDSReplyCode.Ok.getCode())`?  Basically fail if anything other than OK is returned?